### ISSUE TITLE
Pool performance, which is even worse than spinningup

### ISSUE THREAD

**Issue by stoneML123** [schatty/d4pg-pytorch#5]
Hello, I compare the result you report in the README with the spinningup benchmark, your d4pg and d3pg are much worse than spinningup in the HalfCheetah-v2 env, maybe you should improve your implementation?

**Comment by schatty** [comment]
Does spinning up have d4pg implementation? If so, I would have a look if I have free time. I checked ddpg plots and it seems they look alike for 100k env steps

**Comment by stoneML123** [comment]
Indeed I ran your experiment for 3M steps, but Its performance is not so good, stops improvement after 100K steps

**Comment by schatty** [comment]
Interesting, if I remember correctly D4PG paper uses many agents, and I tested the code only on a small amount (4 or so) while writing the code (because I had poor cpu at that time), how many agents did you set?

**Comment by stoneML123** [comment]
I also use four agents, indeed there is a bug when you compute N-step:
`state_0, action_0, reward_0 = self.exp_buffer.popleft()
                    discounted_reward = reward_0
                    gamma = self.config['discount_rate']
                    for (_, _, r_i) in self.exp_buffer:
                        discounted_reward += r_i * gamma
                        gamma *= self.config['discount_rate']`

the second line should be modified into `discounted_reward = 1.0`, otherwise, you will compute 
gamma * (r + v_next) instead of (r + gamma * v_next).
No offence, your code is quite elegant, I love it very much, but I really want to figure out why it can't outpeform Spinningup benchmark, and I am reproducing the result based on your code, It is really helpful to get your feedback, thank you !

**Comment by schatty** [comment]
Thanks for pointing out the bug, I will correct it. Regarding the unexpected low results I note that d4pg paper uses 64 parallel agents and probably has slightly different ratio of incoming env transitions per performed train ops. Therefore it is expected that in a long run something may go wrong for the current implementation. When I implemented it a couple of years ago I had small computational resources and didn't check it on many agents. Also the ddpg itself is pretty unstable algorithm and I didn't tune it much

I will try to refresh the code at the end of the next week for the OpenAI envs to have ok results. The current code is far from perfect and I would like to update it if I have free time. Also you are welcome to open PRs to improve the code

**Comment by stoneML123** [comment]
Hello, I found some possible problems and I fixed them on my implementations.

1. # Ignore the "done" signal if it comes from hitting the time
        # horizon (that is, when it's an artificial terminal signal
        # that isn't based on the agent's state)
        d = False if ep_len == max_ep_len else d
if you don't do this, when keep training, value loss will infinitly grow, which makes training very unstable

2.you mat use torch.set_num_threads(1) for each worker, and torch.set_num_threads(args.learner_thread), this could avoid slowing down by torch threads, which speed up a lot.

3. copy.deepcopy(policy_net_cpu) in agent_worker process, because when you use torch.multiprocessing, you are sharing the policy_net_cpu in all subprocesses, which may cause problems when updating params.

4. replay_queue = mp.Queue(maxsize=1024) instead of 64, many datapoints will lose if you use a small queue, also `step = update_step.value
        logger.scalar_summary("data_struct/global_episode", global_episode.value, step)
        logger.scalar_summary("data_struct/replay_queue", n, step)
        logger.scalar_summary("data_struct/batch_queue", batch_queue.qsize(), step)
        logger.scalar_summary("data_struct/replay_buffer", len(replay_buffer), step)`
you may use `logger.scalar_summary("data_struct/replay_queue", n, step)` instead of `logger.scalar_summary("data_struct/replay_queue", replay_queue.qsize(), step)` for true visualization

**Comment by gianiaco** [comment]
Hello everybody.
I was wondering if @stoneML123 has opened a PR regarding improvements he pointed out.
Anyway, great work!

**Comment by schatty** [comment]
I'm closing this issue for now as performance for HalfCheetah-v2 was improved, feel free to open new ones or create PR regarding concrete issues!
